### PR TITLE
Feat pauli string from-to other language

### PR DESCRIPTION
**Pull request by JulienCalistoTD**

### CODE REVIEW

File: mpqp/core/instruction/measurement/pauli_string.py [hunk 1]
```diff
@@ -341,58 +348,134 @@ def _get_dimension_cirq_pauli(cls, pauli: Union[CirqPauliSum, CirqPauliString]):
...
+        for index, atom in enumerate(pauli.op):
+            monomial[pauli.qbits[index]] = PAULI_ATOM_DICT[atom]
+        return PauliStringMonomial(pauli.coeff, monomial)
+
     @classmethod
     def from_other_language(
         cls,
         pauli: Union[
-            QiskitPauliString,
-            BraketPauliString,
```
Comment (Henri-ColibrITD) [inline mpqp/core/instruction/measurement/pauli_string.py:390]: `BraketPauliString` isn't correct ?

File: mpqp/core/instruction/measurement/pauli_string.py [hunk 2]
```diff
@@ -762,4 +998,6 @@ def __hash__(self):
 Matrix representation:
 `\begin{pmatrix}1&0\\0&-1\end{pmatrix}`
 """
+
+PAULI_ATOM_DICT = {"I": I, "X": X, "Y": Y, "Z": Z}
```
Comment (Henri-ColibrITD) [inline mpqp/core/instruction/measurement/pauli_string.py:1002]: this seems to be for internal use only, I would put it as a private attribute of the class or something like that

File: mpqp/core/instruction/measurement/pauli_string.py [hunk 3]
```diff
@@ -444,54 +579,50 @@ def to_other_language(
...
             1.000*X(q(0))+1.000*Y(q(1))+1.000*Z(q(2))
+            >>> print(ps.to_other_language(Language.MY_QLM))
+            [Term(_coeff=TNumber(is_abstract=False, type=0, int_p=1, double_p=None, string_p=None, matrix_p=None, serialized_p=None, complex_p=None), op='X', qbits=[0], _do_validity_check=True), Term(_coeff=TNumber(is_abstract=False, type=0, int_p=1, double_p=None, string_p=None, matrix_p=None, serialized_p=None, complex_p=None), op='Y', qbits=[1], _do_validity_check=True), Term(_coeff=TNumber(is_abstract=False, type=0, int_p=1, double_p=None, string_p=None, matrix_p=None, serialized_p=None, complex_p=None), op='Z', qbits=[2], _do_validity_check=True)]
+            >>> print(ps.to_other_language(Language.QISKIT))
+            SparsePauliOp(['IIX', 'IYI', 'ZII'],
+                          coeffs=[1.+0.j, 1.+0.j, 1.+0.j])
+            >>> print(ps.to_other_language(Language.BRAKET))
+            Sum(TensorProduct(X('qubit_count': 1), I('qubit_count': 1), I('qubit_count': 1)), TensorProduct(I('qubit_count': 1), Y('qubit_count': 1), I('qubit_count': 1)), TensorProduct(I('qubit_count': 1), I('qubit_count': 1), Z('qubit_count': 1)))
```
Comment (Henri-ColibrITD) [inline mpqp/core/instruction/measurement/pauli_string.py:588]: same story here

**Review by Henri-ColibrITD — CHANGES_REQUESTED** [review]
quelques commentaires à régler mais globalement bon travail !

**Comment by Henri-ColibrITD** [comment]
New bug discovered !
```py
from mpqp.measures import PauliString
from braket.circuits.observables import X as bra_X, Y as bra_Y
PauliString.from_other_language(bra_X() + bra_Y())
```
yields
```py
---------------------------------------------------------------------------
AssertionError                            Traceback (most recent call last)
Input In [6], in <cell line: 1>()
----> 1 PauliString.from_other_language(bra_X()+bra_Y())

File ~\Documents\gits\gh-colibrITD\mpqp\mpqp\core\instruction\measurement\pauli_string.py:528, in PauliString.from_other_language(cls, pauli, min_dimension)
    526     return PauliString._from_qiskit(pauli)
    527 elif isinstance(pauli, BraketObservable):
--> 528     return PauliString._from_braket(pauli)
    529 elif isinstance(pauli, Term):
    530     return PauliString._from_my_qml(pauli)

File ~\Documents\gits\gh-colibrITD\mpqp\mpqp\core\instruction\measurement\pauli_string.py:438, in PauliString._from_braket(cls, pauli)
    436 pauli_str = PauliString()
    437 for tensor_product in pauli.summands:
--> 438     assert isinstance(tensor_product, TensorProduct)
    439     pauli_str += tensor_product_to_pauli_sting(tensor_product)
    440 return pauli_str

AssertionError:
```

File: mpqp/core/instruction/measurement/pauli_string.py (hunk 2 above)
Comment (JulienCalistoTD) [reply mpqp/core/instruction/measurement/pauli_string.py]: done

File: mpqp/core/instruction/measurement/pauli_string.py [hunk 4]
```diff
@@ -444,54 +579,50 @@ def to_other_language(
             >>> ps = X @ I @ I + I @ Y @ I + I @ I @ Z
             >>> print(ps.to_other_language(Language.CIRQ))
             1.000*X(q(0))+1.000*Y(q(1))+1.000*Z(q(2))
+            >>> print(ps.to_other_language(Language.MY_QLM))
+            [Term(_coeff=TNumber(is_abstract=False, type=0, int_p=1, double_p=None, string_p=None, matrix_p=None, serialized_p=None, complex_p=None), op='X', qbits=[0], _do_validity_check=True), Term(_coeff=TNumber(is_abstract=False, type=0, int_p=1, double_p=None, string_p=None, matrix_p=None, serialized_p=None, complex_p=None), op='Y', qbits=[1], _do_validity_check=True), Term(_coeff=TNumber(is_abstract=False, type=0, int_p=1, double_p=None, string_p=None, matrix_p=None, serialized_p=None, complex_p=None), op='Z', qbits=[2], _do_validity_check=True)]
```
[the remark this answers is not in the document]
Comment (JulienCalistoTD) [reply mpqp/core/instruction/measurement/pauli_string.py]: done

File: mpqp/core/instruction/measurement/pauli_string.py (hunk 3 above)
Comment (JulienCalistoTD) [reply mpqp/core/instruction/measurement/pauli_string.py]: done

**Review by Henri-ColibrITD — APPROVED** [review]

File: mpqp/core/instruction/measurement/pauli_string.py [hunk 5]
```diff
@@ -206,6 +212,47 @@ def simplify(self, inplace: bool = False) -> PauliString:
             self._monomials = res.monomials
         return res
 
+    def hard_simplify(self, inplace: bool = False) -> PauliString:
+        """Hard simplifies the PauliString by combining identical terms and removing
```
Comment (hJaffaliColibritd) [inline mpqp/core/instruction/measurement/pauli_string.py:216]: The docstring of this function doesn't tell us what is the difference between ``hard_simplify`` and the already existing ``simplify``.

Is it supposed to remove the identity monomial when other terms are present?

File: mpqp/core/instruction/measurement/pauli_string.py [hunk 6]
```diff
@@ -762,4 +1045,6 @@ def __hash__(self):
 Matrix representation:
 `\begin{pmatrix}1&0\\0&-1\end{pmatrix}`
 """
+
+_pauli_atom_dict = {"I": I, "X": X, "Y": Y, "Z": Z}
```
Comment (hJaffaliColibritd) [inline mpqp/core/instruction/measurement/pauli_string.py:1049]: Shouldn't we put this at the top of the file (since it is used in the _from... methods)?

**Review by hJaffaliColibritd — APPROVED** [review]
Good work in overall, few comments to resolve, I approve the PR

File: mpqp/core/instruction/measurement/pauli_string.py (hunk 5 above)
Comment (JulienCalistoTD) [reply mpqp/core/instruction/measurement/pauli_string.py:216]: done

**Pull request merged**